edit/listing.go

```diff
@@ -343,3 +384,201 @@ func getListing(ed *Editor) *listing {
 		panic("unreachable")
 	}
 }
+
+type listingItem interface {
+	eval.Value
+	Display() ui.Styled
+	Content() string
+	FilterText() string
+}
+
+type listingOptions struct {
+	modeline   string
+	autoCommit bool
+	keepBottom bool
+	ignoreDup  bool
+	ignoreCase bool
+}
+
+type listingItemString struct {
+	eval.String
+}
+
+func (s *listingItemString) Content() string {
+	return s.String.String()
+}
+
+func (s *listingItemString) Display() ui.Styled {
+	return ui.Unstyled(s.String.String())
+}
+
+func (s *listingItemString) FilterText() string {
+	return s.Content()
+}
+
+type listingItemComplex struct {
+	eval.Map
+}
+
+func (c *listingItemComplex) Content() string {
+	key := eval.String("content")
+	if c.Map.HasKey(key) {
+		return c.Map.IndexOne(key).(eval.String).String()
```

Comment:
This will panic if the user supplied a map whose `content` is not a string.

Consider making `listingItemComplex` a struct of fields `content`, `display`, etc.. You can use a `*string` as their types and use `nil` to signal nonexistence.